File: parsers/kotlin/src/main/kotlin/verity/passport/prover/SodParser.kt
package verity.passport.prover

import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.asn1.cms.SignedData
import org.bouncycastle.asn1.cms.Attribute
import org.bouncycastle.asn1.cms.CMSAttributes
import org.bouncycastle.asn1.x509.Certificate
import java.time.Instant

/**
 * Parses raw SOD (Security Object Document) bytes from an ePassport into
 * the internal [SOD] domain model using ASN.1 / CMS structures.
 */
public object SodParser {

    /**
     * Parses the raw SOD bytes (with optional 0x77 length prefix) into a [SOD]
     * containing the encapsulated content, DSC certificate, and signer info.
     */
    public fun parse(rawBytes: ByteArray): SOD {
        val bytes = stripLengthPrefix(rawBytes)

        val contentInfo = try {
            ContentInfo.getInstance(ASN1Sequence.fromByteArray(bytes))
        } catch (e: Exception) {
            throw PassportError.CmsParsingFailed(e.message ?: "Unknown error")
        }

        val signedData = try {
            SignedData.getInstance(contentInfo.content)
        } catch (e: Exception) {
            throw PassportError.CmsParsingFailed(e.message ?: "Failed to parse SignedData")
        }

        val version = signedData.version.intValueExact()

        val digestAlgorithms = signedData.digestAlgorithms.map { algObj ->
            val algId = org.bouncycastle.asn1.x509.AlgorithmIdentifier.getInstance(algObj)
            DigestAlgorithm.fromOid(algId.algorithm.id)
                ?: throw PassportError.Asn1DecodingFailed("Unsupported digest algorithm: ${algId.algorithm.id}")
        }

        val encapContent = parseEncapContent(signedData)
        val certificate = parseCertificate(signedData)
        val signerInfo = parseSignerInfo(signedData)

        return SOD(
            version = version,
            digestAlgorithms = digestAlgorithms,
            encapContentInfo = encapContent,
            signerInfo = signerInfo,
            certificate = certificate,
            bytes = bytes
        )
    }

    private fun stripLengthPrefix(data: ByteArray): ByteArray {
        return if (data.size >= 4 && data[0] == 0x77.toByte() && data[1] == 0x82.toByte()) {
            data.copyOfRange(4, data.size)
        } else {
            data
        }
    }

    private fun parseEncapContent(signedData: SignedData): EContent {
        val encapInfo = signedData.encapContentInfo
        val eContentOctets = encapInfo.content
            ?: throw PassportError.MissingRequiredField("eContent")

        val eContentBytes = eContentOctets.toASN1Primitive().let { prim ->
            when (prim) {
                is ASN1OctetString -> prim.octets
                else -> prim.encoded
            }
        }

        val ldsSecObj = ASN1Sequence.getInstance(ASN1Primitive.fromByteArray(eContentBytes))
        val ldsEnum = ldsSecObj.objects

        val ldsVersion = (ldsEnum.nextElement() as ASN1Integer).intValueExact()
        val hashAlgId = org.bouncycastle.asn1.x509.AlgorithmIdentifier.getInstance(ldsEnum.nextElement())
        val hashAlgorithm = DigestAlgorithm.fromOid(hashAlgId.algorithm.id)
            ?: throw PassportError.Asn1DecodingFailed("Unsupported hash algorithm: ${hashAlgId.algorithm.id}")

        val dgHashSeq = ASN1Sequence.getInstance(ldsEnum.nextElement())
        val dgHashValues = buildMap<Int, ByteArray> {
            for (i in 0 until dgHashSeq.size()) {
                val dgHash = ASN1Sequence.getInstance(dgHashSeq.getObjectAt(i))
                val dgNumber = (dgHash.getObjectAt(0) as ASN1Integer).intValueExact()
                val dgHashValue = (dgHash.getObjectAt(1) as ASN1OctetString).octets
                put(dgNumber, dgHashValue)
            }
        }

        return EContent(
            version = ldsVersion,
            hashAlgorithm = hashAlgorithm,
            dataGroupHashValues = DataGroupHashValues(dgHashValues),
            bytes = eContentBytes
        )
    }

    private fun parseCertificate(signedData: SignedData): DSC {
        val certs = signedData.certificates
            ?: throw PassportError.MissingRequiredField("certificates")

        if (certs.size() == 0) {
            throw PassportError.MissingRequiredField("DSC certificate")
        }

        val certObj = certs.getObjectAt(0)
        val cert = Certificate.getInstance(certObj)
        return DscParser.parse(cert)
    }

    private fun parseSignerInfo(signedData: SignedData): SignerInfo {
        val signerInfos = signedData.signerInfos
        if (signerInfos.size() == 0) {
            throw PassportError.DataNotFound("No SignerInfo found")
        }

        val si = org.bouncycastle.asn1.cms.SignerInfo.getInstance(signerInfos.getObjectAt(0))
        val siVersion = si.version.intValueExact()

        val digestAlg = DigestAlgorithm.fromOid(si.digestAlgorithm.algorithm.id)
            ?: throw PassportError.Asn1DecodingFailed("Unsupported digest algorithm: ${si.digestAlgorithm.algorithm.id}")

        val sigAlgName = SignatureAlgorithmName.fromOid(si.digestEncryptionAlgorithm.algorithm.id)
            ?: throw PassportError.UnsupportedSignatureAlgorithm(si.digestEncryptionAlgorithm.algorithm.id)
        val sigAlgParams = si.digestEncryptionAlgorithm.parameters?.toASN1Primitive()?.encoded
        val signatureAlgorithm = SignatureAlgorithm(sigAlgName, sigAlgParams)

        val signature = si.encryptedDigest.octets

        val signedAttrs = parseSignedAttrs(si)

        return SignerInfo(
            version = siVersion,
            signedAttrs = signedAttrs,
            digestAlgorithm = digestAlg,
            signatureAlgorithm = signatureAlgorithm,
            signature = signature
        )
    }

    private fun parseSignedAttrs(si: org.bouncycastle.asn1.cms.SignerInfo): SignedAttrs {
        val attrs = si.authenticatedAttributes
            ?: throw PassportError.MissingRequiredField("signedAttrs")

        var contentType = ""
        var messageDigest: ByteArray? = null
        var signingTime: Instant? = null

        for (i in 0 until attrs.size()) {
            val attr = Attribute.getInstance(attrs.getObjectAt(i))
            val oid = attr.attrType.id

            when (oid) {
                CMSAttributes.contentType.id -> {
                    val oidVal = ASN1ObjectIdentifier.getInstance(attr.attrValues.getObjectAt(0))
                    contentType = OidRegistry.getName(oidVal.id)
                }
                CMSAttributes.messageDigest.id -> {
                    val raw = attr.attrValues.getObjectAt(0)
                    messageDigest = raw.toASN1Primitive().encoded
                }
                CMSAttributes.signingTime.id -> {
                    val timeObj = attr.attrValues.getObjectAt(0)
                    signingTime = when (val prim = timeObj.toASN1Primitive()) {
                        is ASN1UTCTime -> prim.date.toInstant()
                        is ASN1GeneralizedTime -> prim.date.toInstant()
                        else -> null
                    }
                }
            }
        }

        if (messageDigest == null) {
            throw PassportError.MissingRequiredField("messageDigest")
        }

        val signedAttrsEncoded = ASN1Set.getInstance(attrs.toASN1Primitive()).getEncoded(ASN1Encoding.DER)

        return SignedAttrs(
            contentType = contentType,
            messageDigest = messageDigest,
            signingTime = signingTime,
            bytes = signedAttrsEncoded
        )
    }
}

File: parsers/swift/Sources/PassportProverLib/Parsing/SodParser.swift
import Foundation

/// Parses CMS SignedData (SOD) from ePassport Security Object Document.
public enum SodParser {

    // MARK: - CMS OIDs

    private static let oidSignedData = "1.2.840.113549.1.7.2"
    private static let oidContentType = "1.2.840.113549.1.9.3"
    private static let oidMessageDigest = "1.2.840.113549.1.9.4"
    private static let oidSigningTime = "1.2.840.113549.1.9.5"

    // MARK: - Public API

    /// Parse raw SOD bytes into a structured `SOD` value.
    public static func parse(rawBytes: Data) throws -> SOD {
        let bytes = stripLengthPrefix(rawBytes)

        // ContentInfo: SEQUENCE { contentType OID, [0] content }
        let (outerSeq, _) = try ASN1.parse(bytes)
        guard outerSeq.tag == ASN1.tagSequence else {
            throw PassportError.cmsParsingFailed("Expected SEQUENCE for ContentInfo")
        }

        let contentInfoChildren = try ASN1.parseSequence(outerSeq.data)
        guard contentInfoChildren.count >= 2 else {
            throw PassportError.cmsParsingFailed("ContentInfo must have at least 2 elements")
        }

        let contentTypeOid = ASN1.parseOID(contentInfoChildren[0].data)
        guard contentTypeOid == oidSignedData else {
            throw PassportError.cmsParsingFailed("Expected SignedData OID, got \(contentTypeOid)")
        }

        guard ASN1.isContextTag(contentInfoChildren[1], number: 0) else {
            throw PassportError.cmsParsingFailed("Expected [0] context tag for SignedData content")
        }

        // SignedData SEQUENCE
        let (signedDataSeq, _) = try ASN1.parse(contentInfoChildren[1].data)
        guard signedDataSeq.tag == ASN1.tagSequence else {
            throw PassportError.cmsParsingFailed("Expected SEQUENCE for SignedData")
        }

        let sdChildren = try ASN1.parseSequence(signedDataSeq.data)
        guard sdChildren.count >= 4 else {
            throw PassportError.cmsParsingFailed("SignedData must have at least 4 elements")
        }

        let version = ASN1.parseIntValue(sdChildren[0].data)
        let digestAlgorithms = try parseDigestAlgorithms(sdChildren[1])
        let encapContent = try parseEncapContent(sdChildren[2])

        // Locate certificates [0] and signerInfos SET in remaining children
        var certificate: DSC?
        var signerInfo: SignerInfo?

        for i in 3 ..< sdChildren.count {
            let child = sdChildren[i]
            if ASN1.isContextTag(child, number: 0) {
                certificate = try parseCertificate(child.data)
            } else if child.tag == ASN1.tagSet {
                signerInfo = try parseSignerInfo(child.data)
            }
        }

        guard let cert = certificate else {
            throw PassportError.missingRequiredField("DSC certificate")
        }
        guard let si = signerInfo else {
            throw PassportError.missingRequiredField("SignerInfo")
        }

        return SOD(
            version: version,
            digestAlgorithms: digestAlgorithms,
            encapContentInfo: encapContent,
            signerInfo: si,
            certificate: cert,
            bytes: bytes
        )
    }

    // MARK: - Internal Parsing

    private static func stripLengthPrefix(_ data: Data) -> Data {
        if data.count >= 4, data[data.startIndex] == 0x77, data[data.startIndex + 1] == 0x82 {
            return Data(data[(data.startIndex + 4)...])
        }
        return data
    }

    private static func parseDigestAlgorithms(_ node: ASN1Node) throws -> [DigestAlgorithm] {
        try ASN1.parseSequence(node.data).map { algNode in
            let algChildren = try ASN1.parseSequence(algNode.data)
            guard !algChildren.isEmpty, algChildren[0].tag == ASN1.tagOID else {
                throw PassportError.asn1DecodingFailed("Expected OID in AlgorithmIdentifier")
            }
            let oid = ASN1.parseOID(algChildren[0].data)
            guard let alg = DigestAlgorithm.fromOid(oid) else {
                throw PassportError.asn1DecodingFailed("Unsupported digest algorithm: \(oid)")
            }
            return alg
        }
    }

    private static func parseEncapContent(_ node: ASN1Node) throws -> EContent {
        guard node.tag == ASN1.tagSequence else {
            throw PassportError.cmsParsingFailed("Expected SEQUENCE for EncapsulatedContentInfo")
        }

        let children = try ASN1.parseSequence(node.data)
        guard children.count >= 2, ASN1.isContextTag(children[1], number: 0) else {
            throw PassportError.missingRequiredField("eContent")
        }

        // Inside [0], there's an OCTET STRING containing the LDS SecurityObject
        let (eContentOctet, _) = try ASN1.parse(children[1].data)
        let eContentBytes = eContentOctet.tag == ASN1.tagOctetString ? eContentOctet.data : children[1].data

        // LDS Security Object: SEQUENCE { version, hashAlg, dgHashValues }
        let (ldsSeq, _) = try ASN1.parse(eContentBytes)
        let ldsChildren = try ASN1.parseSequence(ldsSeq.data)
        guard ldsChildren.count >= 3 else {
            throw PassportError.asn1DecodingFailed("LDS Security Object needs at least 3 elements")
        }

        let ldsVersion = ASN1.parseIntValue(ldsChildren[0].data)

        let hashAlgChildren = try ASN1.parseSequence(ldsChildren[1].data)
        guard !hashAlgChildren.isEmpty else {
            throw PassportError.asn1DecodingFailed("Empty hash algorithm identifier")
        }
        let hashOid = ASN1.parseOID(hashAlgChildren[0].data)
        guard let hashAlgorithm = DigestAlgorithm.fromOid(hashOid) else {
            throw PassportError.asn1DecodingFailed("Unsupported hash algorithm: \(hashOid)")
        }

        // Data group hash values
        let dgHashSeqChildren = try ASN1.parseSequence(ldsChildren[2].data)
        var dgHashValues: [Int: Data] = [:]
        for dgHashNode in dgHashSeqChildren {
            let dgFields = try ASN1.parseSequence(dgHashNode.data)
            guard dgFields.count >= 2 else { continue }
            dgHashValues[ASN1.parseIntValue(dgFields[0].data)] = dgFields[1].data
        }

        return EContent(
            version: ldsVersion,
            hashAlgorithm: hashAlgorithm,
            dataGroupHashValues: DataGroupHashValues(values: dgHashValues),
            bytes: eContentBytes
        )
    }

    private static func parseCertificate(_ data: Data) throws -> DSC {
        let certNodes = try ASN1.parseSequence(data)
        guard !certNodes.isEmpty else {
            throw PassportError.missingRequiredField("DSC certificate")
        }
        return try DscParser.parse(certNode: certNodes[0])
    }

    private static func parseSignerInfo(_ data: Data) throws -> SignerInfo {
        let siNodes = try ASN1.parseSequence(data)
        guard !siNodes.isEmpty else {
            throw PassportError.dataNotFound("No SignerInfo found")
        }

        let siChildren = try ASN1.parseSequence(siNodes[0].data)
        guard siChildren.count >= 5 else {
            throw PassportError.cmsParsingFailed("SignerInfo must have at least 5 elements")
        }

        let siVersion = ASN1.parseIntValue(siChildren[0].data)

        // siChildren[1] = signerIdentifier (skip)
        // siChildren[2] = digestAlgorithm
        let digestAlgChildren = try ASN1.parseSequence(siChildren[2].data)
        guard !digestAlgChildren.isEmpty else {
            throw PassportError.asn1DecodingFailed("Empty digest algorithm in SignerInfo")
        }
        let digestOid = ASN1.parseOID(digestAlgChildren[0].data)
        guard let digestAlg = DigestAlgorithm.fromOid(digestOid) else {
            throw PassportError.asn1DecodingFailed("Unsupported digest algorithm: \(digestOid)")
        }

        // Locate signedAttrs [0], signatureAlgorithm, and signature
        var signedAttrsNode: ASN1Node?
        var sigAlgNode: ASN1Node?
        var signatureNode: ASN1Node?

        for idx in 3 ..< siChildren.count {
            let child = siChildren[idx]
            if ASN1.isContextTag(child, number: 0) {
                signedAttrsNode = child
            } else if child.tag == ASN1.tagSequence, sigAlgNode == nil {
                sigAlgNode = child
            } else if child.tag == ASN1.tagOctetString {
                signatureNode = child
            }
        }

        guard let attrsNode = signedAttrsNode else {
            throw PassportError.missingRequiredField("signedAttrs")
        }
        guard let sigAlg = sigAlgNode else {
            throw PassportError.missingRequiredField("signatureAlgorithm in SignerInfo")
        }
        guard let sigNode = signatureNode else {
            throw PassportError.missingRequiredField("signature in SignerInfo")
        }

        let signatureAlgorithm = try DscParser.parseAlgorithmIdentifier(sigAlg)
        let signedAttrs = try parseSignedAttrs(attrsNode)

        return SignerInfo(
            version: siVersion,
            signedAttrs: signedAttrs,
            digestAlgorithm: digestAlg,
            signatureAlgorithm: signatureAlgorithm,
            signature: sigNode.data
        )
    }

    private static func parseSignedAttrs(_ node: ASN1Node) throws -> SignedAttrs {
        let attrNodes = try ASN1.parseSequence(node.data)

        var contentType = ""
        var messageDigest: Data?
        var signingTime: Date?

        for attrNode in attrNodes {
            let attrChildren = try ASN1.parseSequence(attrNode.data)
            guard attrChildren.count >= 2, attrChildren[0].tag == ASN1.tagOID else { continue }

            let oid = ASN1.parseOID(attrChildren[0].data)
            let valuesChildren = try ASN1.parseSequence(attrChildren[1].data)
            guard !valuesChildren.isEmpty else { continue }

            switch oid {
            case oidContentType:
                if valuesChildren[0].tag == ASN1.tagOID {
                    contentType = OidRegistry.getName(ASN1.parseOID(valuesChildren[0].data))
                }
            case oidMessageDigest:
                let rawNode = valuesChildren[0]
                messageDigest = ASN1.encodedBytes(tag: rawNode.tag, content: rawNode.data)
            case oidSigningTime:
                signingTime = ASN1.parseTime(valuesChildren[0])
            default:
                break
            }
        }

        guard let msgDigest = messageDigest else {
            throw PassportError.missingRequiredField("messageDigest")
        }

        // Re-encode signedAttrs as SET for signature verification
        let signedAttrsEncoded = ASN1.reencodeAsSet(node.data)

        return SignedAttrs(
            contentType: contentType,
            messageDigest: msgDigest,
            signingTime: signingTime,
            bytes: signedAttrsEncoded
        )
    }
}
